### ISSUE TITLE
Python Comparison Scripts

### ISSUE THREAD

**Issue by allisonvacanti** [allisonvacanti/nvbench#9]
NVBench has a work-in-progress JSON output format and I'm working on a very basic python script to compare two JSON files.

We should grow this functionality into a more complete set of analysis tools. At minimum, this should cover the features provided by Google Benchmarks' excellent comparison scripts.

If anyone is interested in writing some python to help with this, let me know. I'll update this issue once I have finalized the JSON output format.

**Comment by shwina** [comment]
I'd be interested to help out here!

**Comment by vyasr** [comment]
I can help as well if you need extra hands.

**Comment by allisonvacanti** [comment]
Initial work is in #14.

**Comment by vyasr** [comment]
@allisonvacanti what's next here? #14 helped close the gap, but I don't recall exactly how far it got us or what we still need to do. RAPIDS is making a push to formalize and analyze our benchmarks more, so migrating fully to `nvbench` is probably going to become a priority in the near future and I'm happy to help out in making sure that we have sufficient feature parity with google bench.
CC @shwina in case you want to continue being involved too.

**Comment by robertmaynard** [comment]
Basic thresholding and comparing multiple files was added in #48

**Comment by allisonvacanti** [comment]
There's a lot we *could* still do, such as filtering the results by benchmark name/index and axis values. But I don't think these are essential right now.

Are there any "must have" features for RAPIDS that we're missing?

BTW, I'm working on a branch that makes some changes to the JSON file layout to make things more consistent. I hope to have that merged by the end of the week, time permitting 🤞

**Comment by vyasr** [comment]
I assume the changes you're referring to are #70? It looks great! 🎉 

@robertmaynard @jrhemstad @harrism any thoughts on what we would need to see in nvbench to make the transition from gbench smooth for RAPIDS?

**Comment by jrhemstad** [comment]
The gbench compare script had the ability to do a [U Test](https://github.com/google/benchmark/blob/main/docs/tools.md#u-test) between two samples to determine if there was a statistically significant difference between the populations. Do we have anything like that in nvbench yet? 

It's very helpful when looking at small differences in performance to establish if the difference is just "noise" or actually meaningful.

**Comment by allisonvacanti** [comment]
@vyas Yep! That PR has all of my pending changes to the JSON/Python stuff.

@jrhemstad We don't have anything like that at the moment.

------

One feature that I'd like to see at some point is the ability to compare performance between different benchmarks that use the same axes.

For example, see NVIDIA/thrust#1016, which points out that `thrust::all_of` is slower than `thrust::count_if`. It'd be nice to be able to write some automated tests that check the performance of equivalent algorithms and identify these sorts of issues.
